BitArray::Iterator and RowReader::FindPattern

axxel opened the pull request:
Expand the existing BitArray::Iterator idea and its usage across the
OneD Readers. Details see commit message.

Note: while looking at the possibilities to simplify the code by using std algorithms I noticed the fact that the code uses your BitArray::Iterator only halfheartedly, meaning it was using both int indices and iterators in parallel. One thing led to the next and in the end I spend a lot more time on this than I probably should have ;). I hope you find this PR useful. Let me know if you find issues, I’ll gladly look into them.

huycn commented on core/src/oned/ODCode128Reader.cpp:
```diff
@@ -51,63 +51,31 @@ static const int CODE_START_B = 104;
 
-static DecodeStatus
-DecodeCode(const BitArray& row, std::vector<int>& counters, int rowOffset, int& outCode)
+static bool DecodeCode(std::vector<int>& counters, int* outCode)
```
Same here, assert() is preferred.

huycn commented on core/src/oned/ODITFReader.cpp:
```diff
@@ -70,8 +70,7 @@ static const std::array<std::array<int, 5>, 10> PATTERNS = {
 */
-static DecodeStatus
-DecodeDigit(const std::array<int, 5>& counters, int& outCode)
+static bool DecodeDigit(const std::array<int, 5>& counters, int* outCode)
```
Here also, assert() please.

huycn commented on core/src/oned/ODITFReader.cpp:
```diff
@@ -96,26 +95,22 @@ DecodeDigit(const std::array<int, 5>& counters, int& outCode)
 * @throws NotFoundException if decoding could not complete successfully
 */
-static DecodeStatus DecodeMiddle(const BitArray& row, int payloadStart, int payloadEnd, std::string& resultString)
+static bool DecodeMiddle(BitArray::Iterator begin, BitArray::Iterator end, std::string* resultString)
```
And this last one, assert() please.

axxel commented:
Glad you like it (hence, the work was not a waste of time :)). I added 3 asserts and changed the string result function to remove the pointer altogether in that case.
